src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs

```diff
@@ -31,9 +38,51 @@ public abstract class TextLoader
         /// <exception cref="InvalidDataException"/>
         /// <exception cref="OperationCanceledException"/>
         public virtual Task<TextAndVersion> LoadTextAndVersionAsync(CancellationToken cancellationToken)
+        {
+            if (ImmutableInterlocked.GetOrAdd(ref s_isObsoleteLoadTextAndVersionAsyncOverriden, GetType(), IsObsoleteLoadTextAndVersionAsyncOverriden))
+            {
 #pragma warning disable CS0618 // Type or member is obsolete
-            => LoadTextAndVersionAsync(workspace: null, documentId: null, cancellationToken);
+                return LoadTextAndVersionAsync(workspace: null, documentId: null, cancellationToken);
 #pragma warning restore CS0618
+            }
+
+            throw new NotImplementedException($"{GetType()} must implement LoadTextAndVersionAsync");
+
+            static bool IsObsoleteLoadTextAndVersionAsyncOverriden(Type type)
+            {
+                var subTypes = new TemporaryArray<Type>();
+                while (type != typeof(TextLoader))
+                {
+                    subTypes.Add(type);
```

Comment:
>subTypes.Add(type); [](http://example.com/codeflow?start=20&length=19)

Does it make sense to take advantage of the s_isObsoleteLoadTextAndVersionAsyncOverriden cache for each subtype?